azureeyemodule/app/iot/iot_interface.cpp

```diff
@@ -147,62 +149,104 @@ static inline std::unordered_map<std::string, std::string> parse_onvif_message(c
     size_t pos = 0;
-    while ((pos = dup.find_first_of(delimiters)) != std::string::npos)
+    while ((pos = dup.find_first_of(delimiter1)) != std::string::npos)
     {
         std::string part = dup.substr(0, pos);
         if (part.length() > 0)
         {
-            tokens_and_values.push_back(part);
+            tokens_and_values.push_back(part + "_");
         }
         dup.erase(0, pos + 1);
     }
 
     // There should be one more part at the end
-    tokens_and_values.push_back(dup);
+    tokens_and_values.push_back(dup + "_");
 
     // Now go through the vector of strings and convert it into a map
-    std::string curtok;
+
     std::unordered_map<std::string, std::string> result;
     for (size_t i = 0; i < tokens_and_values.size(); i++)
     {
-        if ((i % 2) == 0)
-        {
-            curtok = tokens_and_values.at(i);
-        }
-        else
+        std::string curtok;
+        int count = 0;
+        pos = 0;
+        while ((pos = tokens_and_values[i].find_first_of(delimiter2)) != std::string::npos)
         {
-            result[curtok] = tokens_and_values.at(i);
+            std::string part = tokens_and_values[i].substr(0, pos);
+            if (part.length() > 0)
+            {
+                switch (count) 
+                {
+                    case(0):
+                        curtok = part;
+                        break;
+                    case(1):
+                        result[curtok] = part;
+                        break;
+                    case(2):
+                    // the only feature by far take in two parameters is snapshot
+                        if (curtok != "snapshot") 
+                        {
+                            util::log_error("Malformed ONVIF control message: " +msg);
+                        }
+                        //result[curtok] = result[curtok] + "/" + part;
+                        result[curtok] = part;
+                        break;
+                    default:
+                        util::log_error("Malformed ONVIF control message: " +msg);
+                }
+            }
+            tokens_and_values[i].erase(0, pos + 1);
+            count += 1;
         }
     }
-
-    // If there are not an even number of items in the vector, the message was malformed.
-    if ((tokens_and_values.size() % 2) != 0)
-    {
-        util::log_error("Malformed ONVIF control message. Will try to work with it: " + msg);
-        result[curtok] = std::string("");
-    }
-
     return result;
 }
 
 /** RECV_MSG message handler. Handles incoming messages for the ONVIF channel. */
 static IOTHUBMESSAGE_DISPOSITION_RESULT receive_onvif_message(IOTHUB_MESSAGE_HANDLE message, void *unused)
 {
-    const std::string msg = std::string(IoTHubMessage_GetMessageId(message));
-    util::log_info("Got an ONVIF control message: " + msg);
-
-    auto deserialized_msg = parse_onvif_message(msg);
+    const unsigned char* buffer = 0;
+    size_t s = 256;
```

Comment:
IoTHubMessage_GetByteArray() will fill in this value for you, so don't assign it to anything, as it is kind of confusing if you do.